main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -289,34 +289,63 @@ object Safety {
   }
 
   /**
-    * Checks that an upcast is sound.
+    * Checks that an upcast is safe.
     *
-    * An upcast is considered sound if:
+    * An upcast is considered safe if:
     *
     * (a) the expression has the exact same flix type
     *
     * (b) the actual expression is a java subtype of the expected java type
     *
+    * (c) the actual expression is a function and a subtype of the expected function.
+    *
     * AND
     *
-    * the purity is being cast from `pure` -> `ef` -> `impure`.
+    * the purity of a function is being cast from `pure` -> `ef` -> `impure`.
     *
-    * @param actual   the expression being upcast.
     * @param expected the upcast expression itself.
+    * @param actual   the expression being upcast.
     */
-  private def isSoundUpcast(actual: Expression, expected: Expression): Boolean = {
-    // check flix types are equal
-    // or java type is subtype of upcast java type
-    // check purity is ok
-    // pure -> ef -> impure
-    // ef -> ef and ef2
-    val types = actual.tpe == expected.tpe
-    val purities = (actual.pur, expected.pur) match {
-      case (Type.Pure, _) => true
-      case (_, Type.Impure) => true
-      case _ => false
-    }
-    types && purities
+  private def isSuperTypeOf(expected: Type, actual: Type, contravariantPos: Boolean = false): Boolean = (expected.typeConstructor, actual.typeConstructor) match {
```

Comment:
A few things:

1. Type association is conventionally calculated according to the subtype relation, rather than the supertype relation. I would recommend instead defining `isSubtypeOf(sub: Type, super: Type): Bool`
2. I don't think you need the `contravariantPos` flag. Instead you can reverse the arguments in the recursive call as necessary.
3. "Contravariant" and "Covariant" are properties of of a type parameter. Regardless of if you are checking for supertype or subtype, `a -> b` is always contravariant in `a` and covariant in `b`.